runtime/3rdparty/ggml/src/ggml.c

```diff
@@ -18956,121 +19417,82 @@ struct ggml_cplan ggml_graph_plan(const struct ggml_cgraph * cgraph, int n_threa
-            state->shared->ec = GGML_STATUS_ABORTED;
-        }
-
-        ggml_barrier(state->shared);
+enum ggml_status ggml_graph_compute(struct ggml_cgraph * cgraph, struct ggml_cplan * cplan) {
+    {
+        GGML_ASSERT(cplan);
+        GGML_ASSERT(cplan->n_threads > 0);
 
-        if (state->shared->ec != GGML_STATUS_SUCCESS) {
-            break;
+        if (cplan->work_size > 0) {
+            GGML_ASSERT(cplan->work_data);
         }
     }
 
-    return 0;
-}
-
-enum ggml_status ggml_graph_compute(struct ggml_cgraph * cgraph, struct ggml_cplan * cplan) {
-    GGML_ASSERT(cplan);
-    GGML_ASSERT(cplan->n_threads > 0);
-    GGML_ASSERT(cplan->work_size == 0 || cplan->work_data != NULL);
-
-    int n_threads = cplan->n_threads;
+    const int n_threads = cplan->n_threads;
 
     struct ggml_compute_state_shared state_shared = {
         /*.cgraph                  =*/ cgraph,
         /*.cgraph_plan             =*/ cplan,
+        /*.perf_node_start_cycles  =*/ 0,
+        /*.perf_node_start_time_us =*/ 0,
         /*.n_threads               =*/ n_threads,
-        /*.n_barrier               =*/ 0,
-        /*.n_barrier_passed        =*/ 0,
+        /*.n_active                =*/ n_threads,
+        /*.node_n                  =*/ -1,
+        /*.node_task               =*/ GGML_TASK_TYPE_FINALIZE,
         /*.abort_callback          =*/ NULL,
         /*.abort_callback_data     =*/ NULL,
-        /*.current_chunk           =*/ 0,
-        /*.ec                      =*/ GGML_STATUS_SUCCESS,
     };
 
-#ifdef GGML_USE_OPENMP
-    if (n_threads > 1) {
-        #pragma omp parallel num_threads(n_threads)
-        {
-            #pragma omp single
-            {
-                // update the number of threads from the actual number of threads that we got from OpenMP
-                n_threads = omp_get_num_threads();
-                state_shared.n_threads = n_threads;
-            }
-
-            struct ggml_compute_state worker = {
-                .thrd   = 0,
-                .ith    = omp_get_thread_num(),
-                .shared = &state_shared,
-            };
-            ggml_graph_compute_thread(&worker);
-        }
-    } else {
-        struct ggml_compute_state worker = {
-            .thrd   = 0,
-            .ith    = 0,
-            .shared = &state_shared,
-        };
-        ggml_graph_compute_thread(&worker);
-    }
-#else
+    counter = 0;
     struct ggml_compute_state * workers = alloca(sizeof(struct ggml_compute_state)*n_threads);
 
-    for (int j = 0; j < n_threads; ++j) {
+    chrome_trace_beg_user(&trace_scope[0], "submit", n_threads, 0);
+    for (int j = 1; j < n_threads; ++j) {
         workers[j] = (struct ggml_compute_state) {
             .thrd   = 0,
             .ith    = j,
             .shared = &state_shared,
+            .ec = GGML_STATUS_SUCCESS,
         };
+    
+        ggml_worker_submit(ggml_graph_compute_thread, &workers[j]);
```

Comment:
1. submit work with ggml_graph_computer_thread
2. transmits the signal to the multi thread.
3. the thread executes ggml_graph_computer_thread after wakeup.